File: suites/osprey-compute-kernels/reference/cases/tak/tak.c
#include <stdio.h>
#include <stdint.h>

static int64_t sub(int64_t a, int64_t b) {
    return a - b;
}

static int64_t tak(int64_t x, int64_t y, int64_t z) {
    if (x > y) {
        return tak(
            tak(sub(x, 1), y, z),
            tak(sub(y, 1), z, x),
            tak(sub(z, 1), x, y));
    }
    return z;
}

int main(void) {
    printf("%lld\n", (long long)tak(32, 16, 8));
    return 0;
}

File: suites/osprey-compute-kernels/reference/cases/tak/tak.rs
fn sub(a: i64, b: i64) -> i64 {
    a - b
}

fn tak(x: i64, y: i64, z: i64) -> i64 {
    if x > y {
        tak(
            tak(sub(x, 1), y, z),
            tak(sub(y, 1), z, x),
            tak(sub(z, 1), x, y),
        )
    } else {
        z
    }
}

fn main() {
    println!("{}", tak(32, 16, 8));
}
